### PR TITLE
Use pickle protocol 5 with NumPy object arrays

### PR DESCRIPTION
**Pull request by jakirkham**
Depends on PR ( https://github.com/dask/distributed/pull/3868 )

Builds off the work in PR ( https://github.com/dask/distributed/pull/3784 ) to support pickle protocol 5. Supports out-of-band pickling when serializing NumPy object arrays. This can be useful when serializing ragged arrays for example or really any NumPy object arrays containing data that supports out-of-band pickling.

### CODE REVIEW

File: distributed/protocol/numpy.py [hunk 1]
```diff
@@ -23,7 +23,10 @@ def itemsize(dt):
 def serialize_numpy_ndarray(x):
     if x.dtype.hasobject:
         header = {"pickle": True}
-        frames = [pickle.dumps(x)]
+        frames = [None]
+        buffer_callback = lambda f: frames.append(memoryview(f))
```
Comment (mrocklin) [inline distributed/protocol/numpy.py:27]: Why are we converting into a `memoryview` here?

Comment (jakirkham) [reply distributed/protocol/numpy.py:26]: The `buffer_callback` is called on values of type `PickleBuffer`. Here's an example:

```python
In [1]: import pickle                                                           

In [2]: import numpy as np                                                      

In [3]: a = np.array([np.arange(3), np.arange(4, 6)], dtype=object)             

In [4]: l = []; d = pickle.dumps(a, protocol=5, buffer_callback=l.append)       

In [5]: l                                                                       
Out[5]: [<pickle.PickleBuffer at 0x115a64740>, <pickle.PickleBuffer at 0x115a647c0>]
```

Though `PickleBuffer` is a wrapper that expects to be passed some [`bytes`-like object]( https://docs.python.org/3.8/glossary.html#term-bytes-like-object ), which it wraps zero-copy. In the example above these are the nested NumPy arrays.

However as the rest of the serialization code expects frames to be `bytes` or `memoryview`s. We would either need to retool it to support `PickleBuffer` (with special casing when it is not available) or we can simply convert it to a `memoryview`, which is also zero-copy.

We've gone for the latter path of converting to a `memoryview` similar to what was done in PR ( https://github.com/dask/distributed/pull/3784 ). Note this is done without loss of generality.

**Comment by jakirkham** [comment]
This should be ready on my end. Please let me know if anything else is needed 🙂

**Comment by mrocklin** [comment]
One more dumb question (sorry for drawing this out): how does this work if the user doesn't have a more modern version of Python?  If pickle5 isn't installed and we're using an older version of Python then what happens here?

**Comment by jakirkham** [comment]
Not at all. The discussion is welcome. 🙂 Was more noting this is not in draft status any more.

In the case pickle protocol 5 is not supported, we would [skip collecting buffers]( https://github.com/dask/distributed/blob/8480b4fde57d16b27b3079f25b95e61b862b79ae/distributed/protocol/pickle.py#L37-L38 ) when calling `pickle.dumps` or `cloudpickle.dumps` and would [skip calling the `buffer_callback`]( https://github.com/dask/distributed/blob/8480b4fde57d16b27b3079f25b95e61b862b79ae/distributed/protocol/pickle.py#L56-L58 ) ([as there would be no `buffers`]( https://github.com/dask/distributed/blob/8480b4fde57d16b27b3079f25b95e61b862b79ae/distributed/protocol/pickle.py#L35 )). As a result we would have [only one frame here]( https://github.com/dask/distributed/blob/e07d26ba8136703a5896ebffd18167b0a980eaf1/distributed/protocol/numpy.py#L27 ), which results from calling `pickle.dumps` or `cloudpickle.dumps` as before. When it comes to unpickle the data, we would have [the first frame as an argument to `loads`]( https://github.com/dask/distributed/blob/e07d26ba8136703a5896ebffd18167b0a980eaf1/distributed/protocol/numpy.py#L103 ) and an empty `list`/`tuple` for `buffers` so would [just calling `pickle.loads` without `buffers`]( https://github.com/dask/distributed/blob/8480b4fde57d16b27b3079f25b95e61b862b79ae/distributed/protocol/pickle.py#L67 ).

TL;DR it behaves the same as the code does before this change.

**Comment by mrocklin** [comment]
Ah, ok I was thrown by seeing `pickle.dumps(...)` without realizing that this was *our* wrapping around pickle.  Grand.  Thanks for the detailed explanation @jakirkham .

**Comment by jakirkham** [comment]
Yeah can see how that would be confusing. Yep, we are [importing it here]( https://github.com/jakirkham/distributed/blob/e07d26ba8136703a5896ebffd18167b0a980eaf1/distributed/protocol/numpy.py#L5 ).

Of course! Thanks for the review 😄
